lib/improver/spotdata/neighbour_finding.py

```diff
@@ -0,0 +1,445 @@
+        by checking that the nearest grid point barring any other conditions
+        is a land point. If a site is a sea point it will use the nearest
+        neighbour as there should be no vertical displacement difference with
+        other sea points.
+
+        Args:
+        -----
+        cube           : Iris cube of gridded data.
+        sites          : Dictionary of site data, including lat/lon and
+                         altitude information.
+        relative_z     : Sets the preferred vertical displacement of the grid
+                         point relative to the site; above/below/None.
+        land_constraint: A boolean that determines if land sites should only
+                         select from grid points also over land.
+        ancillary_data : A dictionary containing additional model data that
+                         is needed.
+                         Must contain {'orography': <cube of orography>}.
+                         Needs {'land': <cube of land mask>} if using land
+                         constraint.
+
+        Returns:
+        --------
+        neighbours: Numpy array of grid i,j coordinates that are nearest to
+                    each site coordinate given. Includes height difference
+                    between site and returned grid point.
+
+        '''
+        # Use the default nearest neighbour list as a starting point, and
+        # if for some reason it is missing, recreate the list using the fast
+        # method.
+        if default_neighbours is None:
+            neighbour_list = self.fast_nearest_neighbour(cube, sites,
+                                                         ancillary_data)
+        else:
+            neighbour_list = default_neighbours
+
+        orography = data_from_ancillary(ancillary_data, 'orography')
+        if land_constraint:
+            land = data_from_ancillary(ancillary_data, 'land')
+
+        for i_site, site in enumerate(sites.itervalues()):
+            altitude = site['altitude']
+
+            i, j = neighbour_list['i'][i_site], neighbour_list['j'][i_site]
+            edgecase = neighbour_list['edge'][i_site]
+
+            node_list = nearest_n_neighbours(i, j, 9)
+            if edgecase:
+                node_list = node_edge_test(node_list, cube)
+
+            if land_constraint:
+                # Check that we are considering a land point and that at least
+                # one neighbouring point is also land. If not no modification
+                # is made to the nearest neighbour coordinates.
+
+                exclude_self = nearest_n_neighbours(i, j, 9, exclude_self=True)
+                if edgecase:
+                    exclude_self = node_edge_test(exclude_self, cube)
+                if not land[i, j] or not any(land[exclude_self]):
+                    continue
+
+                node_list = ConditionalListExtract('not_equal_to').process(
+                    land, node_list, 0)
+
```

Comment:
It would be good to add an inline comment to the lower half of this method to clarify what's happening.